transport/graphsync/graphsync.go

```diff
@@ -84,7 +84,12 @@ func (t *Transport) OpenChannel(ctx context.Context,
 		return err
 	}
 	internalCtx, internalCancel := context.WithCancel(ctx)
+
 	t.dataLock.Lock()
+	// if we have an existing request pending for the channelID, cancel it first.
+	if cancelF, ok := t.contextCancelMap[channelID]; ok {
```

Comment:
non-blocking, something to consider: we may need to add a go routine that monitors transfers and cancels context for inactivity. reason being setting an overall timeout in go-fil-markets may not be a good sign of failure -- a large piece may just take a very long time. The real measure is transfer inactivity. anyway, let's cross that bridge later.

Reply:
I've added this note to https://github.com/filecoin-project/lotus/issues/3878, an issue I've created to track cleanups.